Adding code to create BackendServices

nikhiljindal opened the pull request:
This uses the created health checks and instance groups annotation on the ingress to create the relevant backend service.

Verified by running `mci create` that it creates the backend service as expected.

cc @G-Harmon @madhusudancs @csbell

nikhiljindal commented:
Verified that the tests are passing

```
ok  	github.com/GoogleCloudPlatform/k8s-multicluster-ingress/app/mci/cmd	0.345s
ok  	github.com/GoogleCloudPlatform/k8s-multicluster-ingress/app/mci/pkg/gcp/backendservice	0.219s
?   	github.com/GoogleCloudPlatform/k8s-multicluster-ingress/app/mci/pkg/gcp/cloudinterface	[no test files]
ok  	github.com/GoogleCloudPlatform/k8s-multicluster-ingress/app/mci/pkg/gcp/healthcheck	0.036s
ok  	github.com/GoogleCloudPlatform/k8s-multicluster-ingress/app/mci/pkg/gcp/loadbalancer	0.337s
?   	github.com/GoogleCloudPlatform/k8s-multicluster-ingress/app/mci/pkg/gcp/namer	[no test files]
?   	github.com/GoogleCloudPlatform/k8s-multicluster-ingress/app/mci/pkg/serviceport	[no test files]
```

nikhiljindal commented:
cc @nicksardo @bowei

bowei commented on app/mci/pkg/gcp/backendservice/backendservicesyncer.go:
```diff
@@ -0,0 +1,134 @@
+
+// EnsureBackendService ensures that the required backend service exists.
+// Does nothing if it exists already, else creates a new one.
+func (h *BackendServiceSyncer) EnsureBackendService(lbName string, ports []sp.ServicePort, hcMap healthcheck.HealthChecksMap, npMap NamedPortsMap, igLinks []string) error {
```
why `h`?

nikhiljindal replied:
Renamed to b

bowei commented on app/mci/pkg/gcp/backendservice/backendservicesyncer.go:
```diff
@@ -0,0 +1,134 @@
+		// Backend service with that name exists already. Check if it matches what we want.
+		if backendServiceMatches(desiredBS, existingBS) {
+			// Nothing to do. Desired backend service exists already.
+			fmt.Println("Desired backend service exists already")
```
why not use glog?

nikhiljindal replied:
glog also prints the timestamp, filename and line number. It is great for debug logging but not for user visible messages.
kubectl also uses fmt for user messages and glog for debugging messages.

bowei commented on glide.yaml:
```diff
@@ -1,12 +1,15 @@
-- k8s-multi-cluster-ingress/app/mci/pkg/gcp/namer
-- k8s-multi-cluster-ingress/app/mci/pkg/serviceport
-- k8s-multi-cluster-ingress/app/mci/pkg/gcp/healthcheck
+- github.com/GoogleCloudPlatform/k8s-multicluster-ingress/app/
```
what is with this ignore?

nikhiljindal replied:
Removed

nikhiljindal commented:
Thanks for the review @bowei 
Updated code as per comments

G-Harmon commented on app/mci/pkg/gcp/backendservice/backendservicesyncer.go:
```diff
@@ -0,0 +1,134 @@
+	// Check if backend service already exists.
+	existingBS, err := b.bsp.GetGlobalBackendService(name)
+	if err == nil {
+		fmt.Println("Backend service", name, "exists already. Checking if it matches our desired backend service", name)
```
why print 'name' twice here? Seems redundant at best and misleading at worst.

G-Harmon commented on app/mci/pkg/gcp/backendservice/backendservicesyncer.go:
```diff
@@ -0,0 +1,134 @@
+			fmt.Println("Desired backend service exists already")
+			return nil
+		}
+		// TODO: Require explicit permission from user before doing this.
```
TODO format: include a name/issue

G-Harmon commented on app/mci/pkg/gcp/backendservice/backendservicesyncer.go:
```diff
@@ -0,0 +1,134 @@
+			// We create the backend service with RATE balancing mode and set max rate
+			// per instance to max value so that all requests in a region are sent to
+			// instances in that region.
+			// Setting rps to 1, for example, will round robin requests amongst all
```
s/will/would/ ? You **aren't** setting it to 1...

G-Harmon commented on app/mci/pkg/gcp/backendservice/backendservicesyncer_test.go:
```diff
@@ -0,0 +1,76 @@
+		t.Fatalf("expected nil error, actual: %v", err)
+	}
+	if len(bs.HealthChecks) != 1 || bs.HealthChecks[0] != hcLink {
+		t.Fatalf("unexpected health check in backend service. expected: %s, got: %v", hcLink, bs.HealthChecks)
```
this doesn't need to be Fatalf-- the following check would still be valid to run.

G-Harmon commented on app/mci/pkg/gcp/healthcheck/healthchecksyncer.go:
```diff
@@ -60,24 +60,28 @@ var _ HealthCheckSyncerInterface = &HealthCheckSyncer{}
 
 // EnsureHealthCheck ensures that the required health check exists.
 // Does nothing if it exists already, else creates a new one.
```
update the comment to mention the newly returned HealthChecksMap

G-Harmon replied:
friendly bump

nikhiljindal replied:
Done. There is also a more detailed comment on the interface.

G-Harmon commented on app/mci/pkg/gcp/healthcheck/interfaces.go:
```diff
@@ -16,10 +16,16 @@ package healthcheck
-	EnsureHealthCheck(lbName string, ports []sp.ServicePort) error
+	// EnsureHealthCheck ensures that the required health checks exist.
+	// Returns a map of port number to the health check for that port. The map contains all the ports for which  it was successfully able to ensure a health check.
+	// Incase of no error, the map will contain all the ports from the given array of ports.
```
s/Incase/In case/

nikhiljindal replied:
fixed

G-Harmon commented on app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go:
```diff
@@ -46,32 +49,115 @@ const (
 type LoadBalancerSyncer struct {
 	lbName string
 	hcs    healthcheck.HealthCheckSyncerInterface
+	bss    backendservice.BackendServiceSyncerInterface
```
comment this member?

G-Harmon commented on app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go:
```diff
@@ -46,32 +49,115 @@ const (
 	hcs    healthcheck.HealthCheckSyncerInterface
+	bss    backendservice.BackendServiceSyncerInterface
 	client kubeclient.Interface
+	igp    ingressig.InstanceGroups
```
why igp? Seems like the 'p' isn't necessary. Or perhaps iGrp. (Actually, you set this to gce.GCECloud, so I don't really understand what this member var is for.)

G-Harmon commented on app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go:
```diff
@@ -46,32 +49,115 @@ const (
+		// Keep aggregating errors and return all at the end, rather than giving up on the first error.
+		err = multierror.Append(err, hcErr)
+	}
+	igs, namedPorts, err := l.getIGsAndNamedPorts(ing)
```
you use multierror above, but then you always overwrite 'err' here and in the next if block...

nikhiljindal replied:
Fixed

G-Harmon commented on app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go:
```diff
@@ -46,32 +49,115 @@ const (
+		}
+		var values []InstanceGroupsAnnotationValue
+		if err := json.Unmarshal([]byte(annotationValue), &values); err != nil {
+			return nil, nil, err
```
If you get an error from Unmarshal, it's going to be very low-level. can you add more context with Errorf?

nikhiljindal replied:
Done

G-Harmon commented on app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go:
```diff
@@ -46,32 +49,115 @@ const (
+		}
+		return igs, namedPorts, nil
+	}
+	return nil, nil, nil
```
this looks unreachable. does Go require you to have that?

nikhiljindal replied:
yes :)

G-Harmon commented on app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go:
```diff
@@ -46,32 +49,115 @@ const (
+// Also note that this returns all named ports on the instance group and not just the ones relevant to the given ingress.
+func (l *LoadBalancerSyncer) getIGsAndNamedPorts(ing *v1beta1.Ingress) ([]string, backendservice.NamedPortsMap, error) {
+	// Keep trying until ingress gets the instance group annotation.
+	for try := true; try; time.Sleep(1 * time.Second) {
```
Should we worry about having an exponential backoff?

nikhiljindal replied:
I dont think so (could be convinced otherwise).

User will be waiting and we need to keep printing the trying message at constant rate to ensure that the tool is working and is not dead.

G-Harmon replied:
I don't think we need to do anything now- add a TODO that we might need backoff here?

If we want to keep the user updated, we can always print how long we're waiting before trying again. There shouldn't be a ton of these outstanding requests to the API server such that it causes a problem.

nikhiljindal replied:
Added the TODO.
Note that there will always be only one outstanding request. We wait for the request to complete and return before trying again.

G-Harmon commented on app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go:
```diff
@@ -80,25 +109,67 @@ func TestCreateLoadBalancer(t *testing.T) {
-	if len(actions) != 1 {
+	// Verify that the client should have been used to fetch service and ingress.
+	if len(actions) != 2 {
 		t.Fatalf("unexpected actions: %v, expected 1 get", actions)
```
fatalf text needs updating

nikhiljindal replied:
done

G-Harmon commented on app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go:
```diff
@@ -80,25 +109,67 @@ func TestCreateLoadBalancer(t *testing.T) {
 		return true, ret, nil
 	})
+	// Return a fake ingress with the instance groups annotation.
+	// Return a fake ingress with the instance groups annotation.
```
repeated comment line?

nikhiljindal replied:
removed

G-Harmon commented on app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go:
```diff
@@ -80,25 +109,67 @@ func TestCreateLoadBalancer(t *testing.T) {
+	}
+	bs := fbs.EnsuredBackendServices[0]
+	if bs.LBName != lbName {
+		t.Fatalf("unexpected lb name in backend service. expected: %s, got: %s", lbName, bs.LBName)
```
These fatals can probably also be Errorfs

nikhiljindal replied:
done

nikhiljindal commented on app/mci/pkg/gcp/backendservice/backendservicesyncer.go:
```diff
@@ -0,0 +1,134 @@
+
+type BackendServiceSyncer struct {
+	namer *utilsnamer.Namer
+	bsp   ingressbs.BackendServices
```
Added a comment.
Backend Services Provider. Similar to HealthChecksProvider in healthchecksyncer.go

nikhiljindal commented on app/mci/pkg/gcp/backendservice/backendservicesyncer.go:
```diff
@@ -0,0 +1,134 @@
+	return err
+}
+
+func (b *BackendServiceSyncer) ensureBackendService(lbName string, port sp.ServicePort, hc *compute.HealthCheck, np *compute.NamedPort, igLinks []string) error {
```
Done

nikhiljindal commented on app/mci/pkg/gcp/backendservice/backendservicesyncer.go:
```diff
@@ -0,0 +1,134 @@
+func (b *BackendServiceSyncer) ensureBackendService(lbName string, port sp.ServicePort, hc *compute.HealthCheck, np *compute.NamedPort, igLinks []string) error {
+	fmt.Println("Ensuring backend service for port:", port)
+	if hc == nil {
+		return fmt.Errorf("missing health check")
```
Added more details.
Note that the parent method already adds details around the corresponding port.

nikhiljindal commented:
Thanks for the review @G-Harmon 
Updated code as per comments.
Also rebased the code to include latest merges on HEAD. Verified that tests still pass :)
PTAL.

nikhiljindal commented:
Thanks Greg. Updated code as per comments.
Verified that all tests are still passing :)

Merging
